<a href="https://colab.research.google.com/github/albertaleksa/stock-markets-analytics-zoomcamp/blob/main/homework1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Stock Market Analytics Zoomcamp**

Module 1 Homework (2026 cohort)

In this homework, we're going to download financial data from various sources and perform simple calculations and analysis.

**Question 1: [Index] S&P 500 Stocks Added to the Index**

**Which year had the highest number of additions (starting from 2020)?**

Using the list of S&P 500 companies from Wikipedia's S&P 500 companies page (https://en.wikipedia.org/wiki/List_of_S%26P_500_companies), download the data including the year each company was added to the index.

Hint: you can use pandas.read_html (https://pandas.pydata.org/docs/reference/api/pandas.read_html.html) to scrape the data into a DataFrame. You may need to add headers to make it work:

```
url = 'https://en.wikipedia.org/wiki/List_of_S%26P_500_companies'
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'}

# Fetch the HTML content with headers
response = requests.get(url, headers=headers)
```

Steps:

1. Create a DataFrame with company tickers, names, and the year they were added.
2. Extract the year from the addition date and calculate the number of stocks added each year.
3. Which (full) year had the highest number of additions, starting from 2020?

▎ Note: When stocks are added to the S&P 500, they usually experience a price bump as investors and index funds buy shares following the announcement.

▎ Context:

"Following the announcement, all four new entrants saw their stock prices rise in extended trading on Friday" - recent examples of S&P 500 additions include DASH, WSM, EXE, TKO in 2025 (Nasdaq article - https://www.nasdaq.com/articles/sp-500-reshuffle-dash-tko-expe-wsm-join-worth-buying).

*Additional*: How many current S&P 500 stocks have been in the index for more than 20 years?

In [16]:
# IMPORTS
import pandas as pd
import requests
from datetime import datetime

In [30]:
url = 'https://en.wikipedia.org/wiki/List_of_S%26P_500_companies'
# 1. Define custom headers to mimic a real web browser
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'}

# Fetch the HTML content with headers
response = requests.get(url, headers=headers)

# 3. Pass the HTML text string to pandas
# read_html always returns a LIST of DataFrames (one for each <table> found)
tables = pd.read_html(response.text)

# 4. Extract the specific table you want (e.g., the first table)
sp500 = tables[0]
sp500 = sp500[["Symbol", "Security", "Date added"]]

# 1. Convert column to datetime
sp500['Date added'] = pd.to_datetime(sp500['Date added'], errors="coerce")

# 2. Extract the year
sp500['Year added'] = sp500['Date added'].dt.year

sp500.head()


/tmp/ipykernel_1278/1067961245.py:10: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(response.text)
/tmp/ipykernel_1278/1067961245.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sp500['Date added'] = pd.to_datetime(sp500['Date added'], errors="coerce")
/tmp/ipykernel_1278/1067961245.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sp500['Year added'] = s

,Symbol,Security,Date added,Year added
0,MMM,3M,1957-03-04,1957
1,AOS,A. O. Smith,2017-07-26,2017
2,ABT,Abbott Laboratories,1957-03-04,1957
3,ABBV,AbbVie,2012-12-31,2012
4,ACN,Accenture,2011-07-06,2011


In [32]:
stocks_since_2020 = sp500[sp500["Year added"] >= 2020]

stocks_since_2020["Year added"].value_counts().sort_index()

,count
Year added,
2020,10
2021,10
2022,15
2023,15
2024,16
2025,18
2026,13


In [29]:
stocks_since_2020["Year added"].value_counts().idxmax()

np.int32(2025)

**Answer**:
2025

In [35]:
today = pd.Timestamp.today().normalize()
cutoff = today - pd.DateOffset(years=20)

more_than_20_years = sp500[sp500["Date added"] < cutoff]

len(more_than_20_years)


224

**Answer**: 224

**Question 2. [Macro] Indexes YTD (as of 21 August 2026)**

**How many indexes (out of 10) have better year-to-date returns than the US (S&P 500) as of August 21, 2026?**

Using Yahoo Finance World Indices data, compare the year-to-date (YTD) performance (1 January-21 August 2026, use the closing price to calculate the growth) of major stock market indexes for the following countries:

- United States - S&P 500 (^GSPC)
- China - Shanghai Composite (000001.SS)
- Hong Kong - HANG SENG INDEX (^HSI)
- Australia - S&P/ASX 200 (^AXJO)
- India - Nifty 50 (^NSEI)
- Canada - S&P/TSX Composite (^GSPTSE)
- Germany - DAX (^GDAXI)
- United Kingdom - FTSE 100 (^FTSE)
- Japan - Nikkei 225 (^N225)
- Mexico - IPC Mexico (^MXX)
- Brazil - Ibovespa (^BVSP)

*Hint*: use start_date='2026-01-01' and end_date='2026-08-21' parameters when downloading daily data in yfinance

Context:

"Global Valuations: Who's Cheap, Who's Not?" (https://simplywall.st/article/beyond-the-us-global-markets-after-yet-another-tariff-update) article suggests "Other regions may be growing faster than the US and you need to diversify."

Reference: Yahoo Finance World Indices - https://finance.yahoo.com/world-indices/

*Additional*: How many of these indexes have better returns than the S&P 500 over 3, 5, and 10 year periods? Do you see the same trend?

Note: For simplicity, ignore currency conversion effects.

In [36]:
# install the main library YFinance
!pip install yfinance

In [37]:
#Fin Data Sources
import yfinance as yf
import pandas_datareader as pdr

In [38]:
start_date = "2026-01-01"
end_date = "2026-08-21"

In [39]:
ticker_obj = yf.Ticker("^GSPC")

sp500 = ticker_obj.history(
    start=start_date,
    end=end_date,
    interval="1d",
    auto_adjust=False
)

sp500.head()

,Open,High,Low,Close,Adj Close,Volume,Dividends,Stock Splits
Date,,,,,,,,
2026-01-02 00:00:00-05:00,6878.109863,6894.870117,6824.310059,6858.470215,6858.470215,4184120000,0.0,0.0
2026-01-05 00:00:00-05:00,6892.189941,6920.379883,6891.560059,6902.049805,6902.049805,5771930000,0.0,0.0
2026-01-06 00:00:00-05:00,6908.029785,6948.689941,6904.020020,6944.819824,6944.819824,5509680000,0.0,0.0
2026-01-07 00:00:00-05:00,6945.069824,6965.689941,6919.189941,6920.930176,6920.930176,5214480000,0.0,0.0
2026-01-08 00:00:00-05:00,6914.109863,6931.279785,6899.330078,6921.459961,6921.459961,5333200000,0.0,0.0


In [45]:
sp500.tail()

,Open,High,Low,Close,Adj Close,Volume,Dividends,Stock Splits
Date,,,,,,,,
2026-08-14 00:00:00-04:00,7806.600098,7810.009766,7776.310059,7785.759766,7785.759766,4159900000,0.0,0.0
2026-08-17 00:00:00-04:00,7790.680176,7790.680176,7744.879883,7745.060059,7745.060059,4452630000,0.0,0.0
2026-08-18 00:00:00-04:00,7700.040039,7713.950195,7688.629883,7691.759766,7691.759766,4602700000,0.0,0.0
2026-08-19 00:00:00-04:00,7716.740234,7743.930176,7700.069824,7707.979980,7707.979980,5184100000,0.0,0.0
2026-08-20 00:00:00-04:00,7690.490234,7699.959961,7639.009766,7641.160156,7641.160156,4759610000,0.0,0.0


In [44]:
first_close = sp500["Close"].iloc[0]
print(first_close)
last_close = sp500["Close"].iloc[-1]
print(last_close)

ytd_return = (last_close / first_close - 1) * 100
print(ytd_return)

6858.47021484375
7641.16015625
11.412019253393835


In [47]:
indexes = {
    "United States": "^GSPC",
    "China": "000001.SS",
    "Hong Kong": "^HSI",
    "Australia": "^AXJO",
    "India": "^NSEI",
    "Canada": "^GSPTSE",
    "Germany": "^GDAXI",
    "United Kingdom": "^FTSE",
    "Japan": "^N225",
    "Mexico": "^MXX",
    "Brazil": "^BVSP"
}

results = {}

for country, ticker in indexes.items():
    ticker_obj = yf.Ticker(ticker)

    history = ticker_obj.history(
        start=start_date,
        end=end_date,
        interval="1d",
        auto_adjust=False
    )

    first_close = history["Close"].iloc[0]
    last_close = history["Close"].iloc[-1]

    ytd_return = (last_close / first_close - 1) * 100

    results[country] = ytd_return

In [48]:
results

{'United States': np.float64(11.412019253393835),
 'China': np.float64(-2.974984823951776),
 'Hong Kong': np.float64(-2.4298315417153993),
 'Australia': np.float64(4.0789203231815785),
 'India': np.float64(-7.322958916814581),
 'Canada': np.float64(14.0574656152199),
 'Germany': np.float64(5.883203166599049),
 'United Kingdom': np.float64(8.010175932582818),
 'Japan': np.float64(27.750744826533214),
 'Mexico': np.float64(0.32497191871372344),
 'Brazil': np.float64(4.601997022530346)}

In [49]:
returns = pd.Series(results).sort_values(ascending=False)

returns

,0
Japan,27.750745
Canada,14.057466
United States,11.412019
United Kingdom,8.010176
Germany,5.883203
Brazil,4.601997
Australia,4.078920
Mexico,0.324972
Hong Kong,-2.429832
China,-2.974985


In [50]:
sp500_return = returns["United States"]

better_than_sp500 = returns[
    returns > sp500_return
].drop("United States", errors="ignore")

len(better_than_sp500)

2

**Answer**: 2

**Question 3. [Index] S&P 500 Market Corrections Analysis**

**Calculate the median drawdown (in %) of significant market corrections in the S&P 500 index.**

For this task, define a correction as an event when a stock index goes down by **at least 5%** from the most recent all-time high.

Steps:

1. Download S&P 500 historical data (period from 1950 to present) using yfinance (daily stats)
2. Identify all-time high closing price points (where price exceeds all previous days' prices)
3. For each pair of consecutive all-time highs, find the minimum price in between
4. Calculate drawdown percentages: (high - low) / high × 100
5. Filter for corrections with at least 5% drawdown
6. Calculate the duration in days for each correction period
7. Determine the 25th, 50th (median), and 75th percentiles for correction durations and drawdowns

Context:

- Investors often wonder about the typical length of market corrections when deciding "when to buy the dip" (Reddit discussion - https://www.reddit.com/r/investing/comments/1jrqnte/when_are_you_buying_the_dip/?rdt=64135).
- A Wealth of Common Sense - How Often Should You Expect a Stock Market Correction? - https://awealthofcommonsense.com/2022/01/how-often-should-you-expect-a-stock-market-correction/


*Hint (use this data to compare with your results)*: Here is the list of top 10 largest corrections by drawdown:

- 2007-10-09 to 2009-03-09: 56.8% drawdown over 517 days
- 2000-03-24 to 2002-10-09: 49.1% drawdown over 929 days
- 1973-01-11 to 1974-10-03: 48.2% drawdown over 630 days
- 1968-11-29 to 1970-05-26: 36.1% drawdown over 543 days
- 2020-02-19 to 2020-03-23: 33.9% drawdown over 33 days
- 1987-08-25 to 1987-12-04: 33.5% drawdown over 101 days
- 1961-12-12 to 1962-06-26: 28.0% drawdown over 196 days
- 1980-11-28 to 1982-08-12: 27.1% drawdown over 622 days
- 2022-01-03 to 2022-10-12: 25.4% drawdown over 282 days
- 1966-02-09 to 1966-10-07: 22.2% drawdown over 240 days

**Question 4. [Stocks] Earnings Surprise Analysis for Amazon (AMZN)**

**Calculate the median 2-day percentage change in stock prices following positive earnings surprise days.**

Steps:

1. Load earnings data for the last few years using the yfinance method get_earnings_dates():

```
import yfinance as yf

ticker = 'AMZN'
ticker_obj = yf.Ticker(ticker)
result = ticker_obj.get_earnings_dates()
```

You should have 25 entries starting from 2020-10-29.

Note: the 1 future earnings date entry will not have Reported EPS
and Surprise % filled in.

2. Download complete historical price data using yfinance
3. Calculate 2-day percentage changes for all historical dates: for each sequence of 3 consecutive trading days (Day 1, Day 2, Day 3), compute the *return* as Close_Day3 / Close_Day1 - 1. (Assume Day 2 may correspond to the earnings announcement.)
4. Filter for positive earnings surprises and calculate the median 2-day return. Then calculate the correlation between the 2-day stock return and the earnings surprise magnitude.

▎ *Hint*: you can generate all correlations at once using pd.corr().

**Context**: Earnings announcements, especially when they exceed analyst expectations, can significantly impact stock prices in the short term.

**Reference**: Yahoo Finance Earnings Calendar -  https://finance.yahoo.com/calendar/earnings?symbol=AMZN

*Additional*: Is there a correlation between the magnitude of the earnings surprise and the stock price reaction? Does the market react differently to earnings surprises during bull vs. bear markets?

**Question 5. [Exploratory, optional] Brainstorm potential idea for your capstone project**

**Free text answer**

Describe the capstone project you would like to pursue, considering your aspirations, ML model predictions, and prior knowledge. Even if you are unsure at this stage, try to generate an idea you would like to explore-such as a specific asset class, country, industry vertical, or investment strategy. Be as specific as possible.

*Example*: I want to build a short-term prediction model for the US/India/Brazil stock markets, focusing on the largest stocks over a 30-day investment horizon. I plan to use RSI and MACD technical indicators and news coverage data to generate predictions.

**Question 6. [Exploratory, optional] Investigate new metrics**

**Free text answer**

Using the data sources we have covered (or any others you find relevant), download and explore a few additional metrics or time series that could be valuable for your project. Briefly explain why you think each metric is useful. This does not need to be a comprehensive list-focus on demonstrating your ability to generate data requests based on your project description, identify and locate the necessary data, and explain how you would retrieve it using Python.

**Submitting the solutions**

Form for submitting: https://courses.datatalks.club/sma-zoomcamp-2026/homework/hw01

**Leaderboard**

Leaderboard link: https://courses.datatalks.club/sma-zoomcamp-2026/leaderboard